# Assignment 2 (DM Project)
### Dataset = Kaggle - Google Play Store Apps (*https://www.kaggle.com/lava18/google-play-store-apps*)

Group 27
1. **Jonathan Vieri** - 219559949
2. **Jason Andrea Limas** - 219559869
3. **Rakyan Satrya Adhikara** - 219548135
4. **Venura Hewage** - 215287247

### Importing all necessary libraries

In [31]:
# Importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import csv

### Inserting the data

In [32]:
# Inserting the data
data = pd.read_csv('data/googleplaystore.csv')
data

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

### Cleaning the data
#### Dropping rows
There are a number of rows that have 'NaN' values and that have duplicate values. Thus, it is better to drop those wors so that the data can be analyzed for better results without any errors.

First of, we drop all the data that have 'NaN' values by using the `dropna()` function.

In [50]:
data.isnull().valye

<bound method DataFrame.any of          App  Category  Rating  Reviews   Size  Installs   Type  Price  \
0      False     False   False    False  False     False  False  False   
1      False     False   False    False  False     False  False  False   
2      False     False   False    False  False     False  False  False   
3      False     False   False    False  False     False  False  False   
4      False     False   False    False  False     False  False  False   
...      ...       ...     ...      ...    ...       ...    ...    ...   
10836  False     False   False    False  False     False  False  False   
10837  False     False   False    False  False     False  False  False   
10838  False     False    True    False  False     False  False  False   
10839  False     False   False    False  False     False  False  False   
10840  False     False   False    False  False     False  False  False   

       Content Rating  Genres  Last Updated  Current Ver  Android Ver  
0       

In [33]:
# Cleans the data using the dropna() and drop_duplicates() function from pandas
# The dropna() function will remove rows which contain missing values
data_clean = data.dropna()
data_clean

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10834     4.0       7                2.6M         500+  Free     0   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10834       Everyone                  Education     June 18, 2017   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10834               1.0.0          4.1 and up  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[9360 rows x 13 columns]

Then, we drop all the columns that have duplicate values by using the `drop_duplicates()` function.

In [34]:
# The drop_duplicates() function will remove any duplicate data
data_clean = data_clean.drop_duplicates()
data_clean

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10834                                      FR Calculator               FAMILY   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10834     4.0       7                2.6M         500+  Free     0   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10834       Everyone                  Education     June 18, 2017   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10834               1.0.0          4.1 and up  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[8886 rows x 13 columns]

After dropping the columns, noticed that the indexes hasn't changed yet. Thus, it is better to reindex the dataframe by using the `reset_index()` function.

In [35]:
# Reindexing the dataframe after dropping rows
# by using reset_index() function
data_clean = data_clean.reset_index(drop=True)
data_clean

App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                   Coloring book moana       ART_AND_DESIGN   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                 Sketch - Draw & Paint       ART_AND_DESIGN   
4                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                 ...                  ...   
8881                                      FR Calculator               FAMILY   
8882                                   Sya9a Maroc - FR               FAMILY   
8883                   Fr. Mike Schmitz Audio Teachings               FAMILY   
8884                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
8885      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating Reviews                Size     Installs  Type Price  \
0        4.1     159                 19M      10,000+  Free     0   
1        3.9     967                 14M     500,000+  Free     0   
2        4.7   87510                8.7M   5,000,000+  Free     0   
3        4.5  215644                 25M  50,000,000+  Free     0   
4        4.3     967                2.8M     100,000+  Free     0   
...      ...     ...                 ...          ...   ...   ...   
8881     4.0       7                2.6M         500+  Free     0   
8882     4.5      38                 53M       5,000+  Free     0   
8883     5.0       4                3.6M         100+  Free     0   
8884     4.5     114  Varies with device       1,000+  Free     0   
8885     4.5  398307                 19M  10,000,000+  Free     0   

     Content Rating                     Genres      Last Updated  \
0          Everyone               Art & Design   January 7, 2018   
1          Everyone  Art & Design;Pretend Play  January 15, 2018   
2          Everyone               Art & Design    August 1, 2018   
3              Teen               Art & Design      June 8, 2018   
4          Everyone    Art & Design;Creativity     June 20, 2018   
...             ...                        ...               ...   
8881       Everyone                  Education     June 18, 2017   
8882       Everyone                  Education     July 25, 2017   
8883       Everyone                  Education      July 6, 2018   
8884     Mature 17+          Books & Reference  January 19, 2015   
8885       Everyone                  Lifestyle     July 25, 2018   

             Current Ver         Android Ver  
0                  1.0.0        4.0.3 and up  
1                  2.0.0        4.0.3 and up  
2                  1.2.4        4.0.3 and up  
3     Varies with device          4.2 and up  
4                    1.1          4.4 and up  
...                  ...                 ...  
8881               1.0.0          4.1 and up  
8882                1.48          4.1 and up  
8883                 1.0          4.1 and up  
8884  Varies with device  Varies with device  
8885  Varies with device  Varies with device  

[8886 rows x 13 columns]

#### Converting values
It is brought to our attention that there are some rows that have wrong datatypes. Therefore, it is better to convert all those necessary columns. Those columns are:
- 'Reviews' (to `int64` datatype)
- 'Price' (to `float` datatype)
- 'Last Updated' (to `datetime` datatype)

In [36]:
# Converting the 'Reviews', 'Price' and 'Last Updated' rows to appropriate datatypes
# 'Reviews' column to int64
data_clean.Reviews = pd.to_numeric(data_clean.Reviews, errors='coerce')
data_clean.Reviews

0          159
1          967
2        87510
3       215644
4          967
         ...  
8881         7
8882        38
8883         4
8884       114
8885    398307
Name: Reviews, Length: 8886, dtype: int64

In [37]:
# 'Price' column to Float
data_clean.Price = data_clean.Price.str.replace('$','')
data_clean.Price = pd.to_numeric(data_clean.Price, errors='coerce')
data_clean.Price

0       0.0
1       0.0
2       0.0
3       0.0
4       0.0
       ... 
8881    0.0
8882    0.0
8883    0.0
8884    0.0
8885    0.0
Name: Price, Length: 8886, dtype: float64

In [38]:
# 'Last Updated' to datetime
data_clean['Last Updated'] = pd.to_datetime(data_clean['Last Updated'])
data_clean['Last Updated']

0      2018-01-07
1      2018-01-15
2      2018-08-01
3      2018-06-08
4      2018-06-20
          ...    
8881   2017-06-18
8882   2017-07-25
8883   2018-07-06
8884   2015-01-19
8885   2018-07-25
Name: Last Updated, Length: 8886, dtype: datetime64[ns]

Columns needed to be dropped:
- `Last Updated`
- `Current Ver`
- `Android Ver`

Key variables:

In [39]:
data_clean = data_clean.drop(['Last Updated', 'Current Ver', 'Android Ver'], 1)
data_clean

App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                   Coloring book moana       ART_AND_DESIGN   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                 Sketch - Draw & Paint       ART_AND_DESIGN   
4                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                 ...                  ...   
8881                                      FR Calculator               FAMILY   
8882                                   Sya9a Maroc - FR               FAMILY   
8883                   Fr. Mike Schmitz Audio Teachings               FAMILY   
8884                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
8885      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating  Reviews                Size     Installs  Type  Price  \
0        4.1      159                 19M      10,000+  Free    0.0   
1        3.9      967                 14M     500,000+  Free    0.0   
2        4.7    87510                8.7M   5,000,000+  Free    0.0   
3        4.5   215644                 25M  50,000,000+  Free    0.0   
4        4.3      967                2.8M     100,000+  Free    0.0   
...      ...      ...                 ...          ...   ...    ...   
8881     4.0        7                2.6M         500+  Free    0.0   
8882     4.5       38                 53M       5,000+  Free    0.0   
8883     5.0        4                3.6M         100+  Free    0.0   
8884     4.5      114  Varies with device       1,000+  Free    0.0   
8885     4.5   398307                 19M  10,000,000+  Free    0.0   

     Content Rating                     Genres  
0          Everyone               Art & Design  
1          Everyone  Art & Design;Pretend Play  
2          Everyone               Art & Design  
3              Teen               Art & Design  
4          Everyone    Art & Design;Creativity  
...             ...                        ...  
8881       Everyone                  Education  
8882       Everyone                  Education  
8883       Everyone                  Education  
8884     Mature 17+          Books & Reference  
8885       Everyone                  Lifestyle  

[8886 rows x 10 columns]

From what we know, there are:
1. 8886 data that will be used
2. 10 types of columns. Among them, there are:
    - 3 Numerical columns
    - 7 Categorical columns

### Print the Numerical data

In [40]:
# Decribe the 'Rating' row
data_clean.Rating.describe().round(5)

count    8886.00000
mean        4.18796
std         0.52243
min         1.00000
25%         4.00000
50%         4.30000
75%         4.50000
max         5.00000
Name: Rating, dtype: float64

In [41]:
# Decribe the 'Rating' row
data_clean.Reviews.describe()

count    8.886000e+03
mean     4.730928e+05
std      2.906007e+06
min      1.000000e+00
25%      1.640000e+02
50%      4.723000e+03
75%      7.131325e+04
max      7.815831e+07
Name: Reviews, dtype: float64

In [42]:
# Decribe the 'Price' row
data_clean.Price.describe()

count    8886.000000
mean        0.963526
std        16.194792
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max       400.000000
Name: Price, dtype: float64

### Print the Categorical Data

In [43]:
# Print first 5 columns of 'Category' row along with count value of it and top 5 'Category'
print(data_clean['Category'].head())
print("Unique members: ", data_clean['Category'].nunique())
df_category = data_clean.groupby('Category')['App'].nunique()
df_category.sort_values(ascending=False).head()

0    ART_AND_DESIGN
1    ART_AND_DESIGN
2    ART_AND_DESIGN
3    ART_AND_DESIGN
4    ART_AND_DESIGN
Name: Category, dtype: object
Unique members:  33


Category
FAMILY       1683
GAME          913
TOOLS         719
FINANCE       302
LIFESTYLE     301
Name: App, dtype: int64

In [44]:
# Print first 5 columns of 'Installs' row along with count value of it and top 5 'Category'
print(data_clean['Installs'].head())
print("Unique members: ", data_clean['Installs'].nunique())
df_installs = data_clean.groupby('Installs')['App'].nunique()
df_installs.sort_values(ascending=False).head()

0        10,000+
1       500,000+
2     5,000,000+
3    50,000,000+
4       100,000+
Name: Installs, dtype: object
Unique members:  19


Installs
1,000,000+     1417
100,000+       1097
10,000+         987
10,000,000+     938
1,000+          696
Name: App, dtype: int64

In [45]:
# Print first 5 columns of 'Size' row along with count value of it and top 5 'Category'
print(data_clean['Size'].head())
print("Unique members: ", data_clean['Size'].nunique())
df_size = data_clean.groupby('Size')['App'].nunique()
df_size.sort_values(ascending=False).head()

0     19M
1     14M
2    8.7M
3     25M
4    2.8M
Name: Size, dtype: object
Unique members:  413


Size
Varies with device    1171
14M                    148
12M                    147
11M                    143
13M                    143
Name: App, dtype: int64

In [46]:
# Print first 5 columns of 'Type' row along with count value of it and top 5 'Category'
print(data_clean['Type'].head())
print("Unique members: ", data_clean['Type'].nunique())
df_type = data_clean.groupby('Type')['App'].nunique()
df_type.sort_values(ascending=False).head()

0    Free
1    Free
2    Free
3    Free
4    Free
Name: Type, dtype: object
Unique members:  2


Type
Free    7590
Paid     602
Name: App, dtype: int64

In [47]:
# Print first 5 columns of 'Content Rating' row along with count value of it and top 5 'Category'
print(data_clean['Content Rating'].head())
print("Unique members: ", data_clean['Content Rating'].nunique())
df_cont_rat = data_clean.groupby('Content Rating')['App'].nunique()
df_cont_rat.sort_values(ascending=False).head()

0    Everyone
1    Everyone
2    Everyone
3        Teen
4    Everyone
Name: Content Rating, dtype: object
Unique members:  6


Content Rating
Everyone           6612
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Name: App, dtype: int64

In [48]:
# Print first 5 columns of 'Genres' row along with count value of it and top 5 'Category'
print(data_clean['Genres'].head())
print("Unique members: ", data_clean['Genres'].nunique())
df_genre = data_clean.groupby('Genres')['App'].nunique()
df_genre.sort_values(ascending=False).head()

0                 Art & Design
1    Art & Design;Pretend Play
2                 Art & Design
3                 Art & Design
4      Art & Design;Creativity
Name: Genres, dtype: object
Unique members:  115


Genres
Tools            718
Entertainment    471
Education        429
Finance          302
Productivity     301
Name: App, dtype: int64

### Correlating data
1. Ratings with Reviews
2. Ratings with Content Rating
3. Installs with Genres
4. Ratings, Reviews and Installs

Low review count w/ high rating -> outliers

Bar chart (**important**) -> Genres, Categories

Box plot (Case the outliers)

Temporary Hypothesis:
- Art & Design have high value on `Installs` and `Reviews`
- Education have high value on `Installs`
- High `Installs` = High `Ratings`
- High `Installs` = High `Reviews`